### PR TITLE
[chip dv] Auto-generated interrupt test

### PR DESCRIPTION
**Pull request by sriyerg**
This is the starting point of the SW interrupt test that will be
auto-generated. Currently, it only caters to UART0. Submitting draft PR
for early feedback, before I add it as a template to the topgen script.

EDIT: In the latest force push, the commit has been updated such that the 
test is now autogenerated using the topgen.py utility. 

Signed-off-by: Srikrishna Iyer <sriyer@google.com>

### CODE REVIEW

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (sriyerg) [inline sw/device/tests/autogen/plic_all_irqs_test.c:63]: @timothytrippel, do you think we could autogen this function for all blocks in the testutils area and a common ISR API that can be used in tests directly?

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (sriyerg) [inline sw/device/tests/autogen/plic_all_irqs_test.c:140]: @timothytrippel, it will be nice to have one function that just enables all interrupts as a DIF.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [reply sw/device/tests/autogen/plic_all_irqs_test.c:63]: Yes absolutely. Created task in #8142 to track this. I will create a separate PR for this.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [reply sw/device/tests/autogen/plic_all_irqs_test.c:140]: As you suggested, just use dif_*_irq_restore_all(<all-1s>).

**Review by weicaiyang — COMMENTED** [review]
This looks great to me. Defer to @timothytrippel for approval.

**Review by cindychip — COMMENTED** [review]
Thanks for the implementation. LGTM. Let us see if SW side has any comments.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [inline sw/device/tests/autogen/plic_all_irqs_test.c:104]: Just a heads up, as a part of the IRQ DIF generation/refactoring an IP's base address will be located directly in the dif_<ip>_t struct, instead of being nested in the dif_<ip>_params_t struct. Also shared result types will be used across most diff functions What this means it that the function prototype:

dif_uart_result_t dif_uart_init(dif_uart_params_t params, dif_uart_t *uart)

will become:

dif_result_t dif_uart_init(mmio_region_t base_addr, dif_uart_t *uart) .

Just an FYI. These changes specifically for the UART will land in #8193. Feel free to wait to merge this on top of #8193, or I can refactor this as a part #8193 if this is merged first. Either works for me!

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [inline sw/device/tests/autogen/plic_all_irqs_test.c:129]: same here

**Comment by timothytrippel** [comment]
A couple comments above, otherwise LGTM. Thanks for doing this!

**Review by timothytrippel — APPROVED** [review]

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (sriyerg) [reply sw/device/tests/autogen/plic_all_irqs_test.c:104]: Ack, thanks. I will wait for your PRs to merge, then I will rebase on top.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [inline sw/device/tests/autogen/plic_all_irqs_test.c:40]: Is there a reason these can't be dif_##_peripheral##_irq_t types throughout? (to avoid casting)

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [inline sw/device/tests/autogen/plic_all_irqs_test.c:167]: (nit) instanances --> instances
(nit) interupt --> interrupt

**Review by timothytrippel — APPROVED** [review]
couple of minor nits, otherwise LGTM!

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (sriyerg) [reply sw/device/tests/autogen/plic_all_irqs_test.c]: It can be, but it will end up getting expanded to 70 variables (2 per IP). That's probably ok though. Updated.

**Comment by sriyerg** [comment]
Updated the test to be fully auto-generated with `topgen` in the latest force-push.

File: sw/device/tests/autogen/meson.build
Comment (timothytrippel) [inline sw/device/tests/autogen/meson.build:34]: nice! good to prep this.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [inline sw/device/tests/autogen/plic_all_irqs_test.c:305]: Just a heads up/FYI, this should be the DIF init function prototype for _most_ IPs, however for IPs that are templated, thus have HW instantiation params (*_params_t), there will be an extra parameter in this function prototype. See #8286 for an example with the alert_handler. Just means that this will have some special cases thats all... square peg, round hole...

**Review by timothytrippel — APPROVED** [review]

**Comment by timothytrippel** [comment]
Updates LGTM, thanks!

**Comment by moidx** [comment]
@mcy can you take a look at this proposal to automate the generation of [chip_plic_all_irqs](https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/data/chip_testplan.hjson?q=chip_plic_all_irqs) chip level tests?

**Comment by sriyerg** [comment]
Added `clang_format()` function to `topgen` in the latest force-push so that the generated C test is formatted and the CI does not complain. Also added warning header in the test indicating how was this test generated. 

Looks like we disable clang-formatting on other auto-generated sources such as top level header. Was there a specific reason to disable it? @cfrantz @moidx @mcy

File: sw/device/tests/autogen/meson.build
Comment (mcy) [inline sw/device/tests/autogen/meson.build:13]: Is the expectation that someone will be manually switching these as they become necessary, or that a script will regenerate this file? Seems like the latter. Probably worth marking all generated files as such so people don't get confused.

File: util/topgen/c_test.py
Comment (mcy) [inline util/topgen/c_test.py:34]: Is there a reason this is split into two classes? From reading this code, it's very hard to understand why it's organized in this manner. These classes just look like configuration data to me?

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (mcy) [inline util/topgen/templates/plic_all_irqs_test.c.tpl:79]: So, the question to ask here: why is this jammed into a generated file, instead of being in a "real" header? It doesn't seem to depend on anything in the template, so maybe it makes sense to inline it into where it's used below, or put it in its own header.

Whatever can be hoisted out of this header probably should.

Also, is there a reason for the leading underscore names? We don't have specfic style guidance here, so we should be consistent with the rest of the project.

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (mcy) [inline util/topgen/templates/plic_all_irqs_test.c.tpl:112]: Throughout: we may want to match the indentation with all the Mako directives and the C in this file.

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (mcy) [inline util/topgen/templates/plic_all_irqs_test.c.tpl:128]: Functions like this one should probably be hoisted into another target, just like the macros. Let's keep things that don't depend on template input separate.

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (mcy) [inline util/topgen/templates/plic_all_irqs_test.c.tpl:107]: Seeing this repeated a lot. I think it may be worthwhile to define a mako `<%def>`? It'd be nice to just write `${comment(p)}` or similar.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [inline sw/device/tests/autogen/plic_all_irqs_test.c:153]: nit: Interurpt -> Interrupt

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [inline sw/device/tests/autogen/plic_all_irqs_test.c:170]: nit: indentation seems to be off here

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [inline sw/device/tests/autogen/plic_all_irqs_test.c:398]: nit: instannces -> instances

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [inline sw/device/tests/autogen/plic_all_irqs_test.c:178]: is there a particular reason this is enclosed in a `do { ... } while (0)` loop? or could we just use a `{ ... }` block instead? (sorry I may be missing something...)

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [inline sw/device/tests/autogen/plic_all_irqs_test.c:494]: does it make sense to call these triggers in pseudo-random order at some point?

File: util/topgen/c_test.py
Comment (msfschaffner) [inline util/topgen/c_test.py:68]: this choice is not standardized. maybe throw an error or warning if more than 2 TL-UL interfaces are present, and none of them is called `core`?

**Review by msfschaffner — COMMENTED** [review]
Thanks for adding this, @sriyerg. I have a couple of small comments / questions.

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (timothytrippel) [reply sw/device/tests/autogen/plic_all_irqs_test.c:178]: I think this is common technique to avoid some gotchas, see here: http://www.bruceblinn.com/linuxinfo/DoWhile.html

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (timothytrippel) [reply util/topgen/templates/plic_all_irqs_test.c.tpl:79]: After thinking about this more from our discussion earlier, I think the style guides are confusing here. Specifically, defining these macros right before they are used, and undefining them immediately after is directly inline with the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros), for which we mostly follow. However, there are a couple sentences in our [(OpenTitan) style guide](​​https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) that seem to toss this guidance out the window, and enable, as @mcy suggests, hoisting the macros to a public header: "_In contrast to the recommendation in the Google C++ style guide, exporting macros as part of the public API is allowed in C code. A typical use case is a header with register definitions._", though the exact use case here is different.

Regardless of my style guide naivete, since this is autogenerated code, I think it makes sense to keep this inlined in the Mako template for loop context you make use of below (as @mcy suggests), and just do away with the macro entirely. This would naturally also remove the leading underscore names :) Likewise with the other macro blocks.

**Comment by sriyerg** [comment]
In the force push 30 minutes ago, I addressed all the pending comments. 
In the latest force push, I added the test to Verilator and FPGA targets.

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (mcy) [reply util/topgen/templates/plic_all_irqs_test.c.tpl:74]: > I did not see any real benefit to moving these to a separate header since they are not going to be used in any other tests, hence no need to make these public.

There is a benefit: they are now real C files with real indexing. The templates are very hard to read, and once we move to Bazel we will probably stop checking generated code in in a lot of cases. Please factor the macros out into a header.

> The leading underscore helps distinguish macro args from the identifiers in the rest of the test code.

I know, but the rest of the project doesn't do this. If you feel strongly, send me a PR to the style guide and fixes to make the entire project consistent with this style. I can see the benefit, but OT is small enough that consistency is important.

File: util/topgen.py
Comment (mcy) [inline util/topgen.py:43]: Todos should be written in more of an "infinitive" tense, and include an associated bug:
```
# TODO(###)
```

"Refactor this in a separate PR" is redundant.

File: util/topgen/templates/meson.build.tpl
Comment (mcy) [inline util/topgen/templates/meson.build.tpl:5]: Why do you need a %text escaping directive here?

File: util/topgen/templates/meson.build.tpl
Comment (mcy) [inline util/topgen/templates/meson.build.tpl:24]: Is it worth indenting all of the directives to make this easier to read?

File: sw/device/tests/autogen/meson.build
Comment (mcy) [reply sw/device/tests/autogen/meson.build:18]: Kinda confused about what the comments are achieving if this isn't human editable. Humans should never look at these files, so why put comments in them?

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (sriyerg) [reply util/topgen/templates/plic_all_irqs_test.c.tpl]: Ok. Moved the macros to headers, and removed leading underscores.

File: sw/device/tests/autogen/meson.build
Comment (sriyerg) [inline sw/device/tests/autogen/meson.build:18]: Indenting Mako directives with inline text makes the indentation in the generated text look horrible. It works for C because `clang-format` but not for other sources. I think we should stick to left-aligned mako code in templates.

File: util/topgen/templates/meson.build.tpl
Comment (mcy) [reply util/topgen/templates/meson.build.tpl:5]: But not for the license? I think there's something wrong here. Is it the multiple ##s?

File: sw/device/tests/autogen/meson.build
Comment (mcy) [reply sw/device/tests/autogen/meson.build]: Right but... the file will get regenerated, right? Like I think we should not be emitting these comments at all.

File: sw/device/tests/plic_all_irqs_test.h
Comment (mcy) [inline sw/device/tests/plic_all_irqs_test.h:9]: Maybe the file should have a name that reflects that? The file name seems to imply that it's an API, rather than being a helper. (E.g., `plic_all_irqs_test_helpers.h`.)

File: sw/device/tests/plic_all_irqs_test.h
Comment (mcy) [inline sw/device/tests/plic_all_irqs_test.h:77]: Hang on, why do you need these no-ops? Presumably Ibex enters the handler before irq_force even returns, so it's not clear what effect these are supposed to have.

Comment (sriyerg) [reply sw/device/tests/plic_all_irqs_test.h]: Ahh yes, I was experimenting with these. They are not needed. Removed.

File: sw/device/tests/plic_all_irqs_test.h
Comment (sriyerg) [reply sw/device/tests/plic_all_irqs_test.h]: Done.

File: util/topgen/templates/meson.build.tpl
Comment (sriyerg) [reply util/topgen/templates/meson.build.tpl]: Fixed.

File: sw/device/tests/autogen/meson.build
Comment (sriyerg) [reply sw/device/tests/autogen/meson.build]: Don't really care either way. Just want this PR merged today, so, whatever it takes. Removed.

File: sw/device/tests/plic_all_irqs_test_helper.h
Comment (timothytrippel) [inline sw/device/tests/plic_all_irqs_test_helper.h:6]: maybe this should be in sw/device/lib/testing/rv_plic_all_irqs_testutils.h ?

**Review by timothytrippel — APPROVED** [review]

File: sw/device/tests/plic_all_irqs_test_helper.h
Comment (sriyerg) [reply sw/device/tests/plic_all_irqs_test_helper.h:6]: If I did, it would imply that other tests can use these macros, which is not the case. Perhaps leave it here?

File: sw/device/tests/autogen/plic_all_irqs_test.c
Comment (msfschaffner) [reply sw/device/tests/autogen/plic_all_irqs_test.c:413]: right, if we mainly want to test connectivity here, this should be good as is.

File: util/topgen/c_test.py
Comment (msfschaffner) [reply util/topgen/c_test.py:70]: ok good, thanks for adding.

**Review by msfschaffner — APPROVED** [review]
LGTM from my side

File: util/topgen/templates/plic_all_irqs_test.c.tpl
Comment (sriyerg) [inline util/topgen/templates/plic_all_irqs_test.c.tpl:106]: Assumption here that all templated IPs are parameterized w.r.t. DIFs.

**Pull request merged**
